File: app/src/main/java/com/jamieadkins/gwent/data/interactor/CardsInteractorFirebase.java
package com.jamieadkins.gwent.data.interactor;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.jamieadkins.gwent.base.BaseSingleObserver;
import com.jamieadkins.gwent.card.CardFilter;
import com.jamieadkins.gwent.data.CardDetails;
import com.jamieadkins.gwent.data.FirebaseUtils;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Callable;

import io.reactivex.Completable;
import io.reactivex.CompletableEmitter;
import io.reactivex.CompletableOnSubscribe;
import io.reactivex.CompletableSource;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.ObservableSource;
import io.reactivex.Single;
import io.reactivex.SingleEmitter;
import io.reactivex.SingleOnSubscribe;
import io.reactivex.SingleSource;
import io.reactivex.schedulers.Schedulers;

/**
 * Deals with firebase.
 */

public class CardsInteractorFirebase implements CardsInteractor {
    private static final String PATCH_PATH = "card-data/latest-patch";

    private final FirebaseDatabase mDatabase = FirebaseDatabase.getInstance();
    private DatabaseReference mCardsReference;
    private final DatabaseReference mMistakesReference;
    private final DatabaseReference mPatchReference;

    private Query mCardsQuery;
    private ValueEventListener mCardListener;

    private static CardsInteractorFirebase mInstance;

    public static CardsInteractorFirebase getInstance() {
        if (mInstance == null) {
            mInstance = new CardsInteractorFirebase();
        }

        return mInstance;
    }

    private CardsInteractorFirebase() {
        mPatchReference = mDatabase.getReference(PATCH_PATH);
        mMistakesReference = mDatabase.getReference("reported-mistakes");
    }

    private void onPatchUpdated(String patch) {
        mCardsReference = mDatabase.getReference("card-data/" + patch);
        // Keep Cards data in cache at all times.
        mCardsReference.keepSynced(true);
    }

    private Single<String> getLatestPatch() {
        return Single.defer(new Callable<SingleSource<? extends String>>() {
            @Override
            public SingleSource<? extends String> call() throws Exception {
                return Single.create(new SingleOnSubscribe<String>() {
                    @Override
                    public void subscribe(final SingleEmitter<String> emitter) throws Exception {
                        ValueEventListener patchListener = new ValueEventListener() {
                            @Override
                            public void onDataChange(DataSnapshot dataSnapshot) {
                                String patch = dataSnapshot.getValue(String.class);
                                onPatchUpdated(patch);
                                emitter.onSuccess(patch);
                            }

                            @Override
                            public void onCancelled(DatabaseError databaseError) {

                            }
                        };

                        mPatchReference.addListenerForSingleValueEvent(patchListener);
                    }
                });
            }
        });
    }

    @Override
    public Observable<RxDatabaseEvent<CardDetails>> getCards(final CardFilter filter) {
        return Observable.defer(new Callable<ObservableSource<? extends RxDatabaseEvent<CardDetails>>>() {
            @Override
            public ObservableSource<? extends RxDatabaseEvent<CardDetails>> call() throws Exception {
                return Observable.create(new ObservableOnSubscribe<RxDatabaseEvent<CardDetails>>() {
                    @Override
                    public void subscribe(final ObservableEmitter<RxDatabaseEvent<CardDetails>> emitter) throws Exception {
                        getLatestPatch()
                                .observeOn(Schedulers.io())
                                .subscribe(new BaseSingleObserver<String>() {
                                    @Override
                                    public void onSuccess(String value) {
                                        mCardsQuery = mCardsReference.orderByChild("name");

                                        if (filter.getSearchQuery() != null) {
                                            String query = filter.getSearchQuery();
                                            int charValue = query.charAt(query.length() - 1);
                                            String endQuery = query.substring(0, query.length() - 1);
                                            endQuery += (char) (charValue + 1);

                                            mCardsQuery = mCardsQuery.
                                                    startAt(query)
                                                    // No 'contains' query so have to fudge it.
                                                    .endAt(endQuery);
                                        }

                                        mCardListener = new ValueEventListener() {
                                            @Override
                                            public void onDataChange(DataSnapshot dataSnapshot) {
                                                emitCards(dataSnapshot, emitter, filter)
                                                        .subscribeOn(Schedulers.io())
                                                        .subscribe();
                                            }

                                            @Override
                                            public void onCancelled(DatabaseError databaseError) {

                                            }
                                        };

                                        mCardsQuery.addListenerForSingleValueEvent(mCardListener);
                                    }
                                });
                    }
                });
            }
        });
    }

    private Completable emitCards(final DataSnapshot cardsSnapshot,
                                  final ObservableEmitter<RxDatabaseEvent<CardDetails>> emitter,
                                  final CardFilter filter) {
        return Completable.defer(new Callable<CompletableSource>() {
            @Override
            public CompletableSource call() throws Exception {
                return Completable.create(new CompletableOnSubscribe() {
                    @Override
                    public void subscribe(CompletableEmitter e) throws Exception {
                        for (DataSnapshot cardSnapshot : cardsSnapshot.getChildren()) {
                            CardDetails cardDetails = cardSnapshot.getValue(CardDetails.class);

                            if (cardDetails == null) {
                                emitter.onError(new Throwable("Card doesn't exist."));
                                emitter.onComplete();
                                return;
                            }

                            // Only add card if the card meets all the filters.
                            // Also check name and info are not null. Those are dud cards.
                            if (filter.doesCardMeetFilter(cardDetails)) {
                                emitter.onNext(
                                        new RxDatabaseEvent<CardDetails>(
                                                cardSnapshot.getKey(),
                                                cardDetails,
                                                RxDatabaseEvent.EventType.ADDED
                                        ));
                            }
                        }

                        emitter.onComplete();
                        e.onComplete();
                    }
                });
            }
        });
    }

    /**
     * We have a separate method for getting one card as it is significantly quicker than using a
     * card filter.
     *
     * @param id id of the card to retrieve
     */
    @Override
    public Single<RxDatabaseEvent<CardDetails>> getCard(final String id) {
        return Single.defer(new Callable<SingleSource<? extends RxDatabaseEvent<CardDetails>>>() {
            @Override
            public SingleSource<? extends RxDatabaseEvent<CardDetails>> call() throws Exception {
                return Single.create(new SingleOnSubscribe<RxDatabaseEvent<CardDetails>>() {
                    @Override
                    public void subscribe(final SingleEmitter<RxDatabaseEvent<CardDetails>> emitter) throws Exception {
                        getLatestPatch()
                                .observeOn(Schedulers.io())
                                .subscribe(new BaseSingleObserver<String>() {
                                    @Override
                                    public void onSuccess(String value) {
                                        mCardsQuery = mCardsReference.child(id);

                                        mCardListener = new ValueEventListener() {
                                            @Override
                                            public void onDataChange(DataSnapshot dataSnapshot) {
                                                CardDetails cardDetails = dataSnapshot.getValue(CardDetails.class);

                                                if (cardDetails == null) {
                                                    emitter.onError(new Throwable("Card doesn't exist."));
                                                    return;
                                                }

                                                emitter.onSuccess(
                                                        new RxDatabaseEvent<CardDetails>(
                                                                dataSnapshot.getKey(),
                                                                cardDetails,
                                                                RxDatabaseEvent.EventType.ADDED
                                                        ));
                                            }

                                            @Override
                                            public void onCancelled(DatabaseError databaseError) {

                                            }
                                        };

                                        mCardsQuery.addListenerForSingleValueEvent(mCardListener);
                                    }
                                });
                    }
                });
            }
        });
    }

    @Override
    public Completable reportMistake(final String cardid, final String description) {
        return Completable.defer(new Callable<CompletableSource>() {
            @Override
            public CompletableSource call() throws Exception {
                return Completable.create(new CompletableOnSubscribe() {
                    @Override
                    public void subscribe(CompletableEmitter emitter) throws Exception {
                        String key = mMistakesReference.push().getKey();
                        Map<String, Object> values = new HashMap<>();
                        values.put("cardId", cardid);
                        values.put("description", description);
                        values.put("fixed", false);

                        Map<String, Object> firebaseUpdates = new HashMap<>();
                        firebaseUpdates.put(key, values);

                        mMistakesReference.updateChildren(firebaseUpdates);

                        emitter.onComplete();
                    }
                });
            }
        });
    }

    @Override
    public void removeListeners() {
        if (mCardsQuery != null) {
            mCardsQuery.removeEventListener(mCardListener);
        }
    }
}

File: app/src/main/java/com/jamieadkins/gwent/data/interactor/CardsInteractorFirebase.kt
package com.jamieadkins.gwent.data.interactor

import com.google.firebase.database.*
import com.jamieadkins.gwent.BuildConfig
import com.jamieadkins.gwent.card.CardFilter
import com.jamieadkins.gwent.data.CardDetails
import com.jamieadkins.gwent.data.FirebaseUtils
import io.reactivex.*
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import java.util.*
import android.content.Context.CONNECTIVITY_SERVICE
import android.util.Log
import com.jamieadkins.gwent.base.BaseSingleObserver
import com.jamieadkins.gwent.data.SearchResult
import kotlin.Comparator
import kotlin.collections.ArrayList


/**
 * Deals with firebase.
 */

class CardsInteractorFirebase private constructor() : CardsInteractor {

    private val mDatabase = FirebaseUtils.getDatabase()
    private var mCardsReference: DatabaseReference? = null
    private val mMistakesReference: DatabaseReference
    private val mPatchReference: DatabaseReference

    private var mCardsQuery: Query? = null
    private var mCardListener: ValueEventListener? = null
    private var mLocale = "en-US"

    init {
        mPatchReference = mDatabase.getReference(PATCH_PATH)
        mPatchReference.keepSynced(true)
        mMistakesReference = mDatabase.getReference("reported-mistakes")
    }

    private fun onPatchUpdated(patch: String?) {
        patch.let {
            mCardsReference = mDatabase.getReference("card-data/" + patch)
            // Keep Cards data in cache at all times.
            mCardsReference?.keepSynced(true)
        }
    }

    private val latestPatch: Single<String>
        get() = Single.defer {
            Single.create(SingleOnSubscribe<String> { emitter ->
                val patchListener = object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        val patch = dataSnapshot.getValue(String::class.java)
                        patch?.let {
                            onPatchUpdated(patch)
                            emitter.onSuccess(patch)
                        }
                    }

                    override fun onCancelled(databaseError: DatabaseError?) {

                    }
                }

                mPatchReference.addListenerForSingleValueEvent(patchListener)
            })
        }

    override fun setLocale(locale: String) {
        mLocale = locale
    }

    override fun getCards(filter: CardFilter): Single<MutableList<CardDetails>> {
        return getCards(filter, null, null, false)
    }

    override fun getCards(filter: CardFilter?, cardIds: List<String>): Single<MutableList<CardDetails>> {
        return getCards(filter, null, cardIds, false)
    }

    override fun getCards(filter: CardFilter?, query: String?, useIntelligentSearch: Boolean): Single<MutableList<CardDetails>> {
        return getCards(filter, query, null, useIntelligentSearch)
    }

    private fun getCards(filter: CardFilter?, query: String?, cardIds: List<String>?, useIntelligentSearch: Boolean): Single<MutableList<CardDetails>> {
        var source: Single<MutableList<CardDetails>> = getCards()

        if (query != null) {
            if (useIntelligentSearch) {
                source = latestPatch.flatMap { patch ->
                    onPatchUpdated(patch)
                    getSearchResult(query, patch).flatMap { result ->
                        val idList = ArrayList<String>()
                        result.value.hits?.forEach {
                            idList.add(it)
                        }

                        getCards(idList)
                    }
                }
            } else {
                source = getCards(query)
            }
        } else if (cardIds != null) {
            source = getCards(cardIds)
        }

        filter?.let {
            source = source.map { cardList ->
                val iterator = cardList.listIterator()
                while (iterator.hasNext()) {
                    val card = iterator.next()
                    if (!filter.doesCardMeetFilter(card)) iterator.remove()
                }
                cardList
            }
        }

        return source
    }

    private fun getCards(): Single<MutableList<CardDetails>> {
        return getCards(null)
    }

    private fun getCards(cardIds: List<String>): Single<MutableList<CardDetails>> {
        val singles: ArrayList<Single<CardDetails>> = ArrayList()
        cardIds.forEach { singles.add(getCard(it)) }
        return Single.merge(singles).toList()
    }

    private fun getCards(query: String?): Single<MutableList<CardDetails>> {
        return latestPatch.flatMap { patch ->
            onPatchUpdated(patch)
            mCardsQuery = mCardsReference!!.orderByChild("name/" + mLocale)

            query?.let {
                val charValue = query[query.length - 1].toInt()
                var endQuery = query.substring(0, query.length - 1)
                endQuery += (charValue + 1).toChar()

                mCardsQuery = mCardsQuery!!.startAt(query)
                        // No 'contains' query so have to fudge it.
                        .endAt(endQuery)
            }
            cardDataSnapshot.flatMap { dataSnapshot ->
                Single.create(SingleOnSubscribe<MutableList<CardDetails>> { emitter ->
                    val cardList = mutableListOf<CardDetails>()
                    for (cardSnapshot in dataSnapshot.children) {
                        val cardDetails = cardSnapshot.getValue(CardDetails::class.java)

                        if (cardDetails == null) {
                            emitter.onError(Throwable("Card doesn't exist."))
                            continue
                        }
                        cardList.add(cardDetails)
                    }

                    emitter.onSuccess(cardList)
                })
                        // IMPORTANT: Firebase forces us back onto UI thread.
                        .subscribeOn(Schedulers.io())
            }
        }
    }

    fun getSearchResult(query: String, patch: String): Single<RxDatabaseEvent<SearchResult>> {
        val key = search(query, patch)
        return Single.defer {
            Single.create(SingleOnSubscribe<RxDatabaseEvent<SearchResult>> { emitter ->
                val searchQuery = mDatabase.getReference("search/$patch/results/$key")

                val searchListener = object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        val result = dataSnapshot.getValue(SearchResult::class.java) ?: return

                        emitter.onSuccess(
                                RxDatabaseEvent(
                                        dataSnapshot.key,
                                        result,
                                        RxDatabaseEvent.EventType.ADDED
                                ))
                    }

                    override fun onCancelled(databaseError: DatabaseError?) {

                    }
                }

                searchQuery.addValueEventListener(searchListener)
            })
        }
    }

    private fun search(query: String, patch: String): String {
        val searchReference = mDatabase.getReference("search/$patch/queries")
        val key = searchReference.push().key
        val firebaseUpdates = HashMap<String, Any>()
        val queryMap = HashMap<String, Any>()
        queryMap.put("query", query)
        firebaseUpdates.put(key, queryMap)

        searchReference.updateChildren(firebaseUpdates)
        return key
    }

    private val cardDataSnapshot: Single<DataSnapshot>
        get() = Single.defer {
            Single.create(SingleOnSubscribe<DataSnapshot> { emitter ->
                mCardListener = object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        emitter.onSuccess(dataSnapshot)
                    }

                    override fun onCancelled(databaseError: DatabaseError?) {

                    }
                }

                mCardsQuery!!.addListenerForSingleValueEvent(mCardListener)
            })
        }

    /**
     * We have a separate method for getting one card as it is significantly quicker than using a
     * card filter.

     * @param id id of the card to retrieve
     */
    override fun getCard(id: String): Single<CardDetails> {
        return latestPatch.flatMap { patch ->
            onPatchUpdated(patch)
            Single.defer {
                Single.create(SingleOnSubscribe<CardDetails> { emitter ->
                    mCardsQuery = mCardsReference!!.child(id)

                    mCardListener = object : ValueEventListener {
                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            val cardDetails = dataSnapshot.getValue(CardDetails::class.java)

                            if (cardDetails == null) {
                                emitter.onError(Throwable("Card doesn't exist."))
                                return
                            }

                            emitter.onSuccess(cardDetails)
                        }

                        override fun onCancelled(databaseError: DatabaseError?) {

                        }
                    }

                    mCardsQuery!!.addListenerForSingleValueEvent(mCardListener)
                })
            }
        }
    }

    override fun reportMistake(cardid: String, description: String): Completable {
        return Completable.defer {
            Completable.create { emitter ->
                val key = mMistakesReference.push().key
                val values = HashMap<String, Any>()
                values.put("cardId", cardid)
                values.put("description", description)
                values.put("fixed", false)

                val firebaseUpdates = HashMap<String, Any>()
                firebaseUpdates.put(key, values)

                mMistakesReference.updateChildren(firebaseUpdates)

                emitter.onComplete()
            }
        }
    }

    override fun removeListeners() {
        if (mCardsQuery != null) {
            mCardsQuery!!.removeEventListener(mCardListener!!)
        }
    }

    companion object {
        private val PATCH_PATH = "patch/" + BuildConfig.CARD_DATA_VERSION
        val instance: CardsInteractorFirebase by lazy { CardsInteractorFirebase() }
    }
}
